src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.processor.combination;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.opensearch.neuralsearch.search.CompoundTopDocs;
+
+/**
+ * Abstracts combination of scores in query search results.
+ */
+@Log4j2
+public class ScoreCombiner {
+
+    private static final Float ZERO_SCORE = 0.0f;
+
+    /**
+     * Performs score combination based on input combination technique. Mutates input object by updating combined scores
+     * Main steps we're doing for combination:
+     *  - create map of normalized scores per doc id
+     *  - using normalized scores create another map of combined scores per doc id
+     *  - count max number of hits among sub-queries
+     *  - sort documents by scores and take first "max number" of docs
+     *  - update query search results with normalized scores
+     *  Different score combination techniques are different in step 2, where we create map of "doc id" - "combined score",
+     *  other steps are same for all techniques.
+     * @param queryTopDocs query results that need to be normalized, mutated by method execution
+     * @param scoreCombinationTechnique exact combination technique that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final List<CompoundTopDocs> queryTopDocs, final ScoreCombinationTechnique scoreCombinationTechnique) {
+        // iterate over results from each shard. Every CompoundTopDocs object has results from
+        // multiple sub queries, doc ids may repeat for each sub query results
+        return queryTopDocs.stream()
+            .map(compoundQueryTopDocs -> combineShardScores(scoreCombinationTechnique, compoundQueryTopDocs))
+            .collect(Collectors.toList());
+    }
+
+    private float combineShardScores(ScoreCombinationTechnique scoreCombinationTechnique, CompoundTopDocs compoundQueryTopDocs) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+            return ZERO_SCORE;
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
```

Comment:
not sure what do you mean, like having a separate class and those methods protected or public. we can make them protected, it works for tests

Reply:
Having a separate class and make those method as public. This way, you can mock all those methods and unit test become simple.

Reply:
making them as `protected` does not make it mockable because they are called internally.

Reply:
I think ```spy()``` will provide partial mock, as we're injecting combiner to the workflow class constructor, if we create instance of combiner with ```spy()``` and then mock some methods they will be mocked

Reply:
With mokito framework, it does not work like that because mokito create proxy outside of class instead of modifying class. Therefore, If you are calling method inside class, it does not call mocked method but just call the method directly.

Reply:
That makes sense. I think we still can test methods individually as class doesn't have state so we passing the input and assert on output or mutated state on the input.

Reply:
Yes. You can test individual method by making it as protected. Individual method test wasn't my concern actually.
My concern was `combineScores` method in which multiple local method is called. If you are okay with writing unit test for that method, I am fine as well.